annotatedcsv/decode.go

```diff
@@ -0,0 +1,463 @@
+// Copyright 2021 InfluxData, Inc. All rights reserved.
+// Use of this source code is governed by MIT
+// license that can be found in the LICENSE file.
+
+package annotatedcsv
+
+import (
+	"encoding/base64"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	ireflect "github.com/influxdata/influxdb-client-go/internal/reflect"
+)
+
+// Decode decodes the current row into x, which should be
+// a pointer to a struct or a pointer to a slice.
+//
+// When it's a pointer to a struct, columns in the row are decoded into
+// appropriate fields in the struct, using the similar tag conventions
+// described by encoding/json to determine how to map
+// column names to struct fields. Tag prefix must start with "csv:":
+//
+//  type Point struct {
+//      Timestamp  time.Time `csv:"_time"`
+//      Value      float32   `csv:"_value"`
+//      Location   string    `csv:"location"`
+//      Sensor     string    `csv:"type"`
+//  }
+//
+//  var p Point
+//  err := r.Decode(&p)
+//
+// When it's a pointer to a slice, the slice is changed
+// to have one element for each column (reusing
+// space in the slice if possible), and each element is
+// set to the value in the corresponding column.
+//
+// When decoding into an empty interface value, the resulting
+// type depends on the column type:
+//
+// - string, tag or unrecognized: string
+// - double: float64
+// - unsignedLong: uint64
+// - long: int64
+// - boolean: bool
+// - duration: time.Duration
+// - dateTime: time.Time
+//
+// Any value can be decoded into a string without
+// error - the result is the value in the CSV, so
+//
+//     var row []string
+//     r.Decode(&row)
+//
+// will always succeed and provide all the values in the column as strings.
+//
+// Similarly, any row can be decoded into []interface{}, which will convert
+// all values into elements of the slice according to the rules for empty interfaces above.
+//
+// Slice capacity will be reused when possible.
+func (r *Reader) Decode(x interface{}) error {
+	t := reflect.TypeOf(x)
+	if err := r.initColumns(t, r.cols); err != nil {
+		return err
+	}
+	return r.decodeRow(reflect.ValueOf(x), r.row)
+}
+
+// initColumns initializes r.colSetters for the given columns.
+// It's called once per section.
```

Comment:
It's called every time `Decode` is called, not once per section.

```suggestion
// initColumns initializes r.colSetters for decoding the given columns
// into a value of the given type.
```